README.md
# koa-hpp2

English | [简体中文](https://github.com/sishen654/koa-hpp2/blob/main/README-zh.md)

## ✨ Introduce

Koa middleware to **protect against HTTP Parameter Pollution attacks**

**Note**：This library is inspired by the Express middleware [hpp](https://www.npmjs.com/package/hpp). On the basis of the original, there are three more modifications:

-   Only the query parameter name added to the whitelist can automatically get the last item of the array (this is the default behavior in the hpp library)
-   If the **checkBodyOnlyForContentType** attribute is set, it must be completely consistent with the `content-type` carried by the request to take effect (the hpp library uses the [type-is](https://www.npmjs.com/package/type-is) library for judgment, and this library is only for simple comparison)
-   Two new properties have been added to the configurable options:
    -   **queryWhitelist**：Separately configure the whitelist of query parameters
    -   **bodyWhitelist**：Separately configure the whitelist of body parameters



## 📦 Install

```bash
npm i koa-hpp2
yarn add koa-hpp2
pnpm add koa-hpp2
```



## 🌍 TS support

```ts
import * as koa from 'koa';
import { Middleware } from 'koa';

declare type HppOption = Partial<{
    checkQuery: Boolean;	// default is true
    checkBody: Boolean;		// default is true
    checkBodyOnlyForContentType: string[] | "*";	// default is *, which means that any request data type is passed
    whitelist: string[] | "*";	// default is *, representing all passed
    queryWhitelist: string[];	// default is []
    bodyWhitelist: string[];	// default is []
}>;
declare function export_default(option?: HppOption): Middleware<koa.DefaultState, koa.DefaultContext, any>;

export { HppOption, export_default as default };
```



## 🔨 Usage

### query filter

```js
import koa from "koa"
import hpp from "koa-hpp2"

const app = new koa()
app.use(hpp({
    queryWhitelist: ["a"]  
    // Only the parameter named a in the query is allowed to pass
    // and the rest of the parameters are placed in the polluted object
}))
// ...
```

```js
// Case 1: Suppose the following request is sent
let data = await axios.get<any, Response>("http://127.0.0.1:3333/", { params: { a: 1, b: 2 } })
// ctx.query = { a: "1" }
// ctx.state.queryPolluted = { b: "2" }
```

```js
// Case 2: carry multiple same parameters
let data = await axios.get<any, Response>("http://127.0.0.1:3333/?a=1&a=222&b=2")
// ctx.query = { a: "222" }
// ctx.state.queryPolluted = { a: ["1", "222"], b: "2" }
```

### body filter

>   **Note**: koa-bodyparser middleware must be added before parsing body

```js
import koa from "koa"
import bodyParser from "koa-bodyparser"
import hpp from "koa-hpp2"

const app = new koa()
app.use(bodyParser())	// Body parsing must be added to filter the body
app.use(hpp({
    bodyWhitelist: ["a"]  
    // Only the parameter named a in the body is allowed to pass
    // and the rest of the parameters are placed in the tainted object
}))
// ...
```

```js
let data = await axios.post<any, Response>("http://127.0.0.1:3333/", { a: 1, b: 2 })
// ctx.request.body = { a: 1 }
// ctx.state.bodyPolluted = { b: 2 }
```

### filter at the same time

```js
import koa from "koa"
import bodyParser from "koa-bodyparser"
import hpp from "koa-hpp2"

const app = new koa()
app.use(bodyParser())
app.use(hpp({
    whitelist: ["a"]  
    // Only the parameter named a in body and query is allowed to pass
    // and the rest of the parameters are placed in the pollution object
}))
// ...
```

```js
let data = await axios.post<any, Response[]>("http://127.0.0.1:3333/?a=1&a=222&b=2", { a: 1, b: 2, c: 3 })
// ctx.query = { a: "222" }
// ctx.state.queryPolluted = { a: ["1", "222"], b: "2", c: "3" }
// ctx.request.body = { a: 1 }
// ctx.state.bodyPolluted = { b: 2, c: 3 }
```



## ⌨️ More usage details

-   This is a middleware, as the name suggests **can be reused multiple times**, and placed anywhere in the code
-   **Multiple attributes can be used together**: such as `whitelist` with `bodyWhitelist` or `queryWhitelist`, you can even use all three together, which will make your application more robust and flexible
-   `checkQuery` and `checkBody` can directly switch on and off the filter
-   `checkBodyOnlyForContentType` this attribute can control the data type of the requested data
-   If you encounter any problems during use, you can submit an issue to the github community or the gitee community













CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.4](https://github.com/sishen654/koa-hpp2/compare/v1.0.2...v1.0.4) (2022-12-19)


### Bug Fixes

* handle promise next ([8f18e8d](https://github.com/sishen654/koa-hpp2/commit/8f18e8da0926ad6a9ae91fd4d08fd27a3920684c))

### [1.0.2](https://github.com/sishen654/koa-hpp2/compare/v1.0.1...v1.0.2) (2022-12-19)

### 1.0.1 (2022-12-19)


### Features

* finish function ([830089a](https://github.com/sishen654/koa-hpp2/commit/830089a59166d5d252480e30efd48711422ce4cb))


__test__/hpp.test.ts
import axios from 'axios'
import mockServer from "./mock-server"
import { afterEach } from 'vitest'

interface Response {
    query: Record<string, any>
    body: Record<string, any>
    queryPolluted: Record<string, any> | undefined
    bodyPolluted: Record<string, any> | undefined
}

axios.interceptors.response.use(res => {
    return res.data
}, err => {
    return Promise.reject(err)
})

afterEach(() => { mockServer.close() })

describe("hpp", () => {
    describe("should validate the options and works", () => {
        it('by turning off the whitelist for invalid type', function () {
            // @ts-ignore
            expect(() => { mockServer.start({ whitelist: ["a", 123, [1, 2, 3]] }) }).toThrowError("[HPP] Please pass only strings into the array. Removed the entry <' + 1 + '>")
            // @ts-ignore
            expect(() => { mockServer.start({ queryWhitelist: [123, [1, 2, 3]] }) }).toThrowError("[HPP] Please pass only strings into the array. Removed the entry <' + 0 + '>")
            // @ts-ignore
            expect(() => { mockServer.start({ bodyWhitelist: [[1, 2, 3]] }) }).toThrowError("[HPP] Please pass only strings into the array. Removed the entry <' + 0 + '>")
        });
    })

    describe("should queryWhitelist option is valid and check ctx.query", () => {
        it("by turning off the checkQuery option", async () => {
            mockServer.start({ queryWhitelist: ["a"], checkQuery: false });
            let data = await axios.get<any, Response>("http://127.0.0.1:3333/", { params: { a: 1, b: 2 } })
            expect(data.query).toEqual({ a: "1", b: "2" });
            expect(data.queryPolluted).toEqual(undefined);
            expect(data.body).toEqual({});
            expect(data.bodyPolluted).toEqual(undefined);
        })
        it("by turning on the checkQuery option", async () => {
            mockServer.start({ queryWhitelist: ["a"] });
            let data = await axios.get<any, Response>("http://127.0.0.1:3333/", { params: { a: 1, b: 2 } })
            expect(data.query).toEqual({ a: "1" });
            expect(data.queryPolluted).toEqual({ b: "2" });
            expect(data.body).toEqual({});
            expect(data.bodyPolluted).toEqual(undefined);
        })
        it("set queryWhitelist and whitelist option at the same time", async () => {
            mockServer.start({ queryWhitelist: ["a"], whitelist: ["b"] });
            let data = await axios.get<any, Response>("http://127.0.0.1:3333/", { params: { a: 1, b: 2, c: 3, d: 4 } })
            expect(data.query).toEqual({ a: "1", b: "2" });
            expect(data.queryPolluted).toEqual({ c: "3", d: "4" });
            expect(data.body).toEqual({});
            expect(data.bodyPolluted).toEqual(undefined);
        })
        it("with two or more same querys", async () => {
            mockServer.start({ queryWhitelist: ["a"] })
            let data = await axios.get<any, Response>("http://127.0.0.1:3333/?a=1&a=222&b=2")
            expect(data.query).toEqual({ a: "222" });
            expect(data.queryPolluted).toEqual({ a: ["1", "222"], b: "2" });
            expect(data.body).toEqual({});
            expect(data.bodyPolluted).toEqual(undefined);
        })
        it("with two or more hpp Middleware", async () => {
            mockServer.start([{ queryWhitelist: ["a", "b", "c"] }, { queryWhitelist: ["a", "b"] }, { queryWhitelist: ["a"] }, { whitelist: [] }])
            let data = await axios.get<any, Response[]>("http://127.0.0.1:3333/?a=1&a=222&b=2&c=3")
            expect(data[0].query).toEqual({ a: "222", b: "2", c: "3" });
            expect(data[0].queryPolluted).toEqual({ a: ["1", "222"] });
            expect(data[0].body).toEqual({});
            expect(data[0].bodyPolluted).toEqual(undefined);

            expect(data[1].query).toEqual({ a: "222", b: "2" });
            expect(data[1].queryPolluted).toEqual({ a: ["1", "222"], c: "3" });
            expect(data[1].body).toEqual({});
            expect(data[1].bodyPolluted).toEqual(undefined);

            expect(data[2].query).toEqual({ a: "222" });
            expect(data[2].queryPolluted).toEqual({ a: ["1", "222"], b: "2", c: "3" });
            expect(data[2].body).toEqual({});
            expect(data[2].bodyPolluted).toEqual(undefined);

            expect(data[3].query).toEqual({});
            expect(data[3].queryPolluted).toEqual({ a: ["1", "222"], b: "2", c: "3" });
            expect(data[3].body).toEqual({});
            expect(data[3].bodyPolluted).toEqual(undefined);
        })
    })

    describe("should bodyWhitelist option is valid and check ctx.request.body", () => {
        it("by turning off the checkBody option", async () => {
            mockServer.start({ bodyWhitelist: ["a"], checkBody: false });
            let data = await axios.post<any, Response>("http://127.0.0.1:3333/", { a: 1, b: 2 })
            expect(data.query).toEqual({});
            expect(data.queryPolluted).toEqual(undefined);
            expect(data.body).toEqual({ a: 1, b: 2 });
            expect(data.bodyPolluted).toEqual(undefined);
        })
        it("by turning on the checkBody option", async () => {
            mockServer.start({ bodyWhitelist: ["a"] });
            let data = await axios.post<any, Response>("http://127.0.0.1:3333/", { a: 1, b: 2 })
            expect(data.query).toEqual({});
            expect(data.queryPolluted).toEqual(undefined);
            expect(data.body).toEqual({ a: 1 });
            expect(data.bodyPolluted).toEqual({ b: 2 });
        })
        it("set bodyWhitelist and whitelist option at the same time", async () => {
            mockServer.start({ bodyWhitelist: ["a"], whitelist: ["b"] });
            let data = await axios.post<any, Response>("http://127.0.0.1:3333/", { a: 1, b: 2, c: 3, d: 4 })
            expect(data.query).toEqual({});
            expect(data.queryPolluted).toEqual(undefined);
            expect(data.body).toEqual({ a: 1, b: 2 });
            expect(data.bodyPolluted).toEqual({ c: 3, d: 4 });
        })
        it("with two or more hpp Middleware", async () => {
            mockServer.start([{ bodyWhitelist: ["a", "b", "c"] }, { bodyWhitelist: ["a", "b"] }, { bodyWhitelist: ["a"] }, { whitelist: [] }])
            let data = await axios.post<any, Response[]>("http://127.0.0.1:3333/", { a: 1, b: 2, c: 3 })
            expect(data[0].body).toEqual({ a: 1, b: 2, c: 3 });
            expect(data[0].bodyPolluted).toEqual({});
            expect(data[0].query).toEqual({});
            expect(data[0].queryPolluted).toEqual(undefined);

            expect(data[1].body).toEqual({ a: 1, b: 2 });
            expect(data[1].bodyPolluted).toEqual({ c: 3 });
            expect(data[1].query).toEqual({});
            expect(data[1].queryPolluted).toEqual(undefined);

            expect(data[2].body).toEqual({ a: 1 });
            expect(data[2].bodyPolluted).toEqual({ b: 2, c: 3 });
            expect(data[2].query).toEqual({});
            expect(data[2].queryPolluted).toEqual(undefined);

            expect(data[3].body).toEqual({});
            expect(data[3].bodyPolluted).toEqual({ a: 1, b: 2, c: 3 });
            expect(data[3].query).toEqual({});
            expect(data[3].queryPolluted).toEqual(undefined);
        })
    })

    describe("should check both", () => {
        it("with unset any option", async () => {
            mockServer.start()
            let data = await axios.post<any, Response>("http://127.0.0.1:3333/", { a: 1, b: 2 }, { params: { a: 1, b: 2 } })
            expect(data.query).toEqual({ a: "1", b: "2" });
            expect(data.queryPolluted).toEqual(undefined);
            expect(data.body).toEqual({ a: 1, b: 2 });
            expect(data.bodyPolluted).toEqual(undefined);
        })
        it('with one and more params', async () => {
            mockServer.start({ whitelist: ["a"] })
            let data = await axios.post<any, Response>("http://127.0.0.1:3333/", { a: 1, b: 2 }, { params: { a: 1, b: 2 } })
            expect(data.query).toEqual({ a: "1" });
            expect(data.queryPolluted).toEqual({ b: "2" });
            expect(data.body).toEqual({ a: 1 });
            expect(data.bodyPolluted).toEqual({ b: 2 });
        })
        it("with two or more hpp Middleware", async () => {
            mockServer.start([{ whitelist: ["a", "b", "c"] }, { whitelist: ["a", "b"] }, { whitelist: ["a"] }, { whitelist: [] }])
            let data = await axios.post<any, Response[]>("http://127.0.0.1:3333/?a=1&a=222&b=2&c=3", { a: 1, b: 2, c: 3 })
            expect(data[0].query).toEqual({ a: "222", b: "2", c: "3" });
            expect(data[0].queryPolluted).toEqual({ a: ["1", "222"] });
            expect(data[0].body).toEqual({ a: 1, b: 2, c: 3 });
            expect(data[0].bodyPolluted).toEqual({});

            expect(data[1].query).toEqual({ a: "222", b: "2" });
            expect(data[1].queryPolluted).toEqual({ a: ["1", "222"], c: "3" });
            expect(data[1].body).toEqual({ a: 1, b: 2 });
            expect(data[1].bodyPolluted).toEqual({ c: 3 });

            expect(data[2].query).toEqual({ a: "222" });
            expect(data[2].queryPolluted).toEqual({ a: ["1", "222"], b: "2", c: "3" });
            expect(data[2].body).toEqual({ a: 1 });
            expect(data[2].bodyPolluted).toEqual({ b: 2, c: 3 });

            expect(data[3].query).toEqual({});
            expect(data[3].queryPolluted).toEqual({ a: ["1", "222"], b: "2", c: "3" });
            expect(data[3].body).toEqual({});
            expect(data[3].bodyPolluted).toEqual({ a: 1, b: 2, c: 3 });
        })
    })
})


gulpfile.js
import gulp from "gulp"
import ts from "gulp-typescript"
import clean from "gulp-clean"
import jsminfy from "gulp-uglify"
import chalk from "chalk"
import merge from "merge2"

const FOLDER = "./src"
const OUT_FOLDER = "./dist"
const TS_FILE_PATH = FOLDER + "/**/*.{ts,tsx}"
const NOT_TS_FILE_PATH = FOLDER + "/**/*.!(*ts|*tsx)"
// ts 配置文件
let tsProject
if (process.env.NODE_ENV === "production") {
  // isolatedModules 打开会导致 .d.ts 文件无法生成
  tsProject = ts.createProject('tsconfig.json', { isolatedModules: false });
} else {
  tsProject = ts.createProject('tsconfig.json');
}

function handlerTs () {
  let tsResult = gulp.src(TS_FILE_PATH).pipe(tsProject())
  if (process.env.NODE_ENV === 'production') {
    // 正式环境中对 dts 进行输出，并对 js 文件进行一个脏压缩处理
    return merge([
      tsResult.dts.pipe(gulp.dest(OUT_FOLDER)),
      tsResult.js.pipe(jsminfy()).pipe(gulp.dest(OUT_FOLDER))
    ]);
  }
  return tsResult.pipe(gulp.dest(OUT_FOLDER))
}

function handlerNotTsFile () {
  return gulp.src(NOT_TS_FILE_PATH)
    .pipe(gulp.dest(OUT_FOLDER))
}

const cleanFolder = function () {
  return gulp.src(OUT_FOLDER, { allowEmpty: true })
    .pipe(clean())
}

const startWatch = function () {
  // 开发环境下开启监听
  if (process.env.NODE_ENV === 'development') {
    gulp.watch(TS_FILE_PATH, handlerTs)
    gulp.watch(NOT_TS_FILE_PATH, handlerNotTsFile)
    console.log(`\n${chalk.bgRed(" watch ")} start watching...\n`);
  } else {
    console.log(`\n${chalk.bgBlue(" success ")} build successful.\n`);
    // 安全退出
    return Promise.resolve()
  }
}

gulp.task('default', gulp.series(cleanFolder, handlerTs, handlerNotTsFile, startWatch));


rollup.config.js
// 如果打包的文件中有引入外部资源时，执行环境没有会导致出现问题，这个包可以将引入的包代码也打包进来
import resolve from '@rollup/plugin-node-resolve';
// rollup.js 编译源码中的模块引用默认只支持 ES6+的模块方式 import/export。
// 然而大量的 npm 模块是基于 CommonJS 模块方式，这就导致了大量 npm 模块不能直接编译使用
// 需要添加这个插件来支持基于 CommonJS 模块方式 npm 包
import commonjs from '@rollup/plugin-commonjs';
// 使用 esbuild 进行打包可以获得更快的打包速度
import esbuild from 'rollup-plugin-esbuild'
// 用于创建 .d.ts 文件，拟补 esbuid 无法生成声明文件问题
import dts from "rollup-plugin-dts";
import path from "node:path"
// 引入 require 函数，方便解析 json 文件
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const tsConfig = require('./tsconfig.json');
const pkg = require('./package.json');

function resolveConfig (option) {
  let { bundlePath = "dist", declaration = true, bundleName, isResolve = false, isCut = false, format = "cjs", isMinify = true, platform = "node", target = "esnext", plugins = [], input, output, esBuildConfig } = option
  // 1 input 必须存在
  if (!input) throw new Error("Input value must valild!")
  // 2 获取 output
  let defaultOutput = {
    file: bundleName ? path.join(bundlePath, bundleName) : undefined,
    dir: bundleName ? undefined : bundlePath,
    format,   // https://rollupjs.org/guide/en/#outputformat
    preserveModules: isCut
  }
  if (output) Object.assign(defaultOutput, output);
  // 3 获取 esBuild 配置
  let esBuildDefaultConfig = {
    minify: process.env.NODE_ENV === 'production' && isMinify,
    platform, // esbuild: export type Platform = 'browser' | 'node' | 'neutral';
    target,   // https://esbuild.github.io/api/#target
  }
  if (esBuildConfig) Object.assign(esBuildDefaultConfig, esBuildConfig);
  // 4 获取 plugins
  let defaultPlugins = [
    esbuild(esBuildDefaultConfig),
    commonjs()
  ]
  if (isResolve) defaultPlugins.push(resolve())
  // 5 创建返回对象
  let returnArr = [{
    input,
    output: defaultOutput,
    watch: {
      clearScreen: true,
      exclude: 'node_modules/**',
      include: 'src/**/*.ts'
    },
    plugins: defaultPlugins.concat(plugins)
  }]
  // 6 判断是否添加声明文件
  if (declaration && tsConfig.compilerOptions.declaration && format === "es") {
    // 初始化生成位置
    if (defaultOutput.file) {
      defaultOutput.dir = path.dirname(defaultOutput.file);
      defaultOutput.file = undefined;
    }
    returnArr.push({
      input,
      output: defaultOutput,
      watch: false,
      plugins: [dts()]
    })
  }
  return returnArr
}

function getBundleName (format, name) {
  let matches = name.match(/\/([\w\W]+)/) || []
  let original = matches[1] ? matches[1] : name
  // 改名
  switch (format) {
    case "es":
      if (path.extname(original) === ".ts") original = original.replace(path.extname(original), ".js");
      else if (path.extname(original) === ".tsx") original = original.replace(path.extname(original), ".jsx");
      return {
        bundleName: original
      }
    case "cjs":
      return {
        bundleName: original.replace(path.extname(original), ".cjs")
      }
    case "amd":
      break;
    case "iife":
      break;
    case "umd":
      break;
    case "system":
      break;
    default:
      return {}
  }
}

function createConfig (configArr = [], format = ["cjs", "es"]) {
  let arr = []
  configArr.forEach(v => {
    // 1 传入字符串
    if (typeof v === "string") {
      format.forEach(t => {
        arr.push(...resolveConfig(Object.assign({ input: v, format: t }, getBundleName(t, v))))
      })
    }
    // 2 传入对象
    else {
      if (v.formatOnce) { format = ["es"] }
      format.forEach(t => {
        if (!t.bundleName) {
          let defaultConfig = Object.assign({ format: t }, getBundleName(t, v.input))
          arr.push(...resolveConfig(Object.assign(defaultConfig, v)))
        } else {
          arr.push(...resolveConfig(Object.assign({ format: t }, v)))
        }
      })
    }
  })
  return arr
}

// { input: "src/cli.ts", declaration: false, formatOnce: true }
export default createConfig(["src/index.ts"])


__test__/mock-server.ts
import koa from "koa"
import bodyParser from "koa-bodyparser"
import multer from "@koa/multer"
import hpp from "../dist/index"
import type { HppOption } from "../src/index"

let server: any = null

export default {
    start(options: HppOption | HppOption[] = {}) {
        const app = new koa()

        // 必须添加 body 解析，才能对 body 进行过滤
        app.use(bodyParser())
        app.use(multer().single())
        if (!Array.isArray(options)) options = [options];
        options.forEach(hppOption => {
            app.use(hpp(hppOption)).use((ctx, next) => {
                // 返回数据
                if (!ctx.state.arr) ctx.state.arr = [];
                // 防止对象引用指向同一个问题
                ctx.state.arr.push(JSON.parse(JSON.stringify({
                    query: ctx.query,
                    body: ctx.request.body,
                    queryPolluted: ctx.state.queryPolluted,
                    bodyPolluted: ctx.state.bodyPolluted
                })))
                next()
            })
        })
        app.use((ctx, next) => {
            // 返回数据
            ctx.body = ctx.state.arr[1] ? ctx.state.arr : ctx.state.arr[0]
            next()
        })
        server = app.listen(3333)
    },
    close() {
        if (server) {
            server.close()
            server = null
        }
    }
}


src/index.ts
import type { ParameterizedContext, Next, Middleware } from "koa"
import typeis from "type-is"

export type HppOption = Partial<{
    checkQuery: Boolean,
    checkBody: Boolean,
    checkBodyOnlyForContentType: string[] | "*",
    whitelist: string[] | "*",
    queryWhitelist: string[],
    bodyWhitelist: string[]
}>

function itemIsString(arr: string[]) {
    arr.forEach((t, i) => {
        if (typeof t !== "string") {
            throw new Error(`[HPP] Please pass only strings into the array. Removed the entry <' + ${i} + '>`)
        }
    })
}

export default function (option: HppOption = {}) {
    let { whitelist = "*", checkQuery = true, checkBody = true, checkBodyOnlyForContentType = "*", queryWhitelist = [], bodyWhitelist = [] } = option
    // 1 判断存在
    if (!whitelist) throw new Error("Please pass either a string or an string array to options.whitelist")

    // 2 判断数组每一项
    if (Array.isArray(whitelist)) { itemIsString(whitelist) }
    itemIsString(queryWhitelist)
    itemIsString(bodyWhitelist)

    // 3 定义过滤函数
    function _filter(type: "query" | "body", ctx: ParameterizedContext, _whitelist: string[]) {
        // @ts-ignore
        let reqData: Record<string, any> = type === "query" ? ctx[type] : ctx.request.body
        let reqPolluted: Record<string, any> = ctx.state[`${type}Polluted`]
        if (!reqPolluted) {
            let obj = {}
            ctx.state[`${type}Polluted`] = obj
            reqPolluted = obj
        }
        // 只有 _whitelist 存在，才会进行过滤
        if (Array.isArray(_whitelist)) {
            // 1 当值为数组，存进污染对象，且原属性的值只取数组最后一位
            _whitelist.forEach(key => {
                let value = reqData[key]
                // http://127.0.0.1:3333/?a=1&a=222&b=2&c=3: req.query = { a: [ '1', '222' ], b: '2', c: '3' }
                if (Array.isArray(value)) {
                    reqPolluted[key] = value
                    reqData[key] = value[value.length - 1]
                }
            })
            // 2 过滤属性
            for (const key in reqData) {
                if (Object.prototype.hasOwnProperty.call(reqData, key)) {
                    // 获取值
                    const value = reqData[key];
                    // 白名单没有该属性，剔除
                    if (!_whitelist.includes(key)) {
                        // 防止重复添加
                        if (!reqPolluted[key]) {
                            reqPolluted[key] = value
                        }
                        delete reqData[key];
                    }
                }
            }
        }
    }

    // 4 返回中间件
    const hpp: Middleware = async function (ctx: ParameterizedContext, next: Next) {
        // 1 通过全部
        if (whitelist === "*" && !queryWhitelist[0] && !bodyWhitelist[0]) return await next();
        // 2 转化为数组
        whitelist = Array.isArray(whitelist) ? whitelist : []
        // 3 过滤
        if (checkQuery && Object.keys(ctx.query)[0]) {
            _filter("query", ctx, whitelist.concat(queryWhitelist))
        }
        // @ts-ignore 需要在前面添加 koa-bodyparser 中间件
        if (checkBody && ctx.request.body && Object.keys(ctx.request.body)[0]) {
            let arr = whitelist.concat(bodyWhitelist)
            if (checkBodyOnlyForContentType === "*") _filter("body", ctx, arr);
            else {
                if (Array.isArray(checkBodyOnlyForContentType) && checkBodyOnlyForContentType.includes(ctx.header["content-type"] as string)) {
                    _filter("body", ctx, arr)
                }
            }
        }
        await next();
    }
    return hpp
}
